plugins/pixart-rf/fu-pxi-firmware.c

```diff
@@ -25,39 +28,57 @@ fu_pxi_rf_firmware_parse (FuFirmware *firmware,
 			  GError **error)
 {
 	gsize bufsz = 0;
-	guint8 pos = 0;
 	const guint8 *buf;
+	guint8 fw_header[PIXART_RF_FW_HEADER_SIZE];
 	g_autoptr(FuFirmwareImage) img = fu_firmware_image_new (fw);
+	const guint8 TAG[PIXART_RF_FW_HEADER_TAG_SIZE] = {0x55, 0xAA, 0x55, 0xAA, 0x55, 0xAA, 0x55, 0xAA};
+	guint8 fw_header_tag[PIXART_RF_FW_HEADER_TAG_SIZE];
+	gboolean check_header_exist = TRUE;
 
 	/* get buf */
 	buf = g_bytes_get_data (fw, &bufsz);
-	if (bufsz < 32) {
+	if (bufsz < sizeof(fw_header)) {
 		g_set_error (error,
 			     G_IO_ERROR,
 			     G_IO_ERROR_FAILED,
 			     "firmware invalid, too small!");
 		return FALSE;
 	}
+	/* get fw header */
+	if (!fu_memcpy_safe (fw_header, sizeof(fw_header), 0x0,
+			     buf, bufsz, bufsz - sizeof(fw_header),
+			     sizeof(fw_header), error)) {
+		g_prefix_error (error, "failed to read fw header ");
+		return FALSE;
+	}
 	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL) {
-		fu_common_dump_raw (G_LOG_DOMAIN, "fw last 32 bytes",
-				    &buf[bufsz - 32], 32);
+		fu_common_dump_raw (G_LOG_DOMAIN, "fw header",
+				    fw_header, sizeof(fw_header));
+	}
+	
+	/* get fw header tag */
+	if (!fu_memcpy_safe (fw_header_tag, sizeof(fw_header_tag), 0x0,
+			     fw_header, sizeof(fw_header), PIXART_RF_FW_HEADER_TAG_OFFSET,
+			     sizeof(fw_header_tag), error)) {
+		g_prefix_error (error, "failed to read fw header tag ");
+		return FALSE;
 	}
 
-	/* find the version tag */
-	for (guint32 idx = 0; idx < 32; idx++) {
-		if (buf[(bufsz - 32) + idx ] == 'v') {
-			pos = idx;
+	/* check the TAG from fw header is correct */
+	for (guint32 idx = 0x0; idx < sizeof(fw_header_tag); idx++) {
+		if (fw_header_tag[idx] != TAG[idx]) {
+			check_header_exist = FALSE;
 			break;
 		}
 	}
 
 	/* set the default version if can not find it in fw bin */
-	if (pos < 32 - 6 && buf[(bufsz - 32) + pos + 1] == '_') {
+	if (check_header_exist) {
 		g_autofree gchar *version = NULL;
 		version = g_strdup_printf ("%c.%c.%c",
-					   buf[(bufsz - 32) + pos + 2],
-					   buf[(bufsz - 32) + pos + 4],
-					   buf[(bufsz - 32) + pos + 6]);
+					   fw_header[0],
+					   fw_header[2],
+					   fw_header[4]);
 		fu_firmware_set_version (firmware, version);
 	} else {
 		fu_firmware_set_version (firmware, "1.0.0");
```

Comment:
Is this really right?  Shouldn't it be something like "0.0.0" if the header didn't exist or parse properly?